mitodl/mit-open
# OCW ETL

# What are the relevant tickets?
Closes #114 
Might also close https://github.com/mitodl/mit-open/issues/121

Related PR: https://github.com/mitodl/salt-ops/pull/1590

# Description (What does it do?)
Refactors old course_catalog code for importing OCW courses to the new `learning_resources` app.



# How can this be tested?
- Run `docker compose build web celery` (new dependency - `retry2`)
- Set the following in your .env file:
  ```
  OCW_BASE_URL=https://ocw.mit.edu/
  OCW_LIVE_BUCKET=ocw-content-live-production
  OCW_WEBHOOK_KEY=key
  ```
- With containers running, post the following JSON to http://localhost:8063/api/v1/ocw_next_webhook/ via postman or curl:
  ```
  {
     "webhook_key": "key",
     "prefix": "courses/esd-04j-frameworks-and-models-in-engineering-systems-engineering-system-design-spring-2007/",
     "version": "live"
  }
  ```

- Check the logs, you should see some lines about that course getting imported.
- Go to http://localhost:8063/api/v1/courses/?platform=ocw, the course above should be returned.  It should have resource content tags, departments, etc.
- Post the following JSON to http://localhost:8063/api/v1/search/, you should get the above course returned:

  ```
  {"post_filter": { "bool": { "must": [{"bool": { "should": [{ "term": {"readable_id": "ESD.04J"}}]}}]}}}
  ```

- Run `docker compose exec web ./manage.py backpopulate_ocw_data --delete --course-name esd-04j-frameworks-and-models-in-engineering-systems-engineering-system-design-spring-2007`, the course/resource should be deleted.
- Wait a minute then post the search query above again, you should not get any results back.
- Run `docker compose exec web ./manage.py backpopulate_ocw_data --delete --course-name esd-04j-frameworks-and-models-in-engineering-systems-engineering-system-design-spring-2007`, the course should be reimported.
- Run the search query again, the course should be returned again.
- You can also try `docker compose exec web ./manage.py backpopulate_ocw_data` to import all ~2500 but it will take a long time.  Or start it let it run awhile, and terminate it whenever you want.


# Additional Context
I added a new library, retry2, to add a common decorator for retrying certain functions that occasionally failed with request errors (to tika or S3) when importing content files
